backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/predicate/impl/LuceneTextTermsPredicate.java

```diff
@@ -0,0 +1,114 @@
+import org.hibernate.search.backend.lucene.search.impl.LuceneSearchValueFieldContext;
+import org.hibernate.search.backend.lucene.search.predicate.impl.AbstractLuceneLeafSingleFieldPredicate;
+import org.hibernate.search.backend.lucene.types.codec.impl.LuceneStandardFieldCodec;
+import org.hibernate.search.engine.search.common.ValueConvert;
+import org.hibernate.search.engine.search.predicate.SearchPredicate;
+import org.hibernate.search.engine.search.predicate.spi.TermsPredicateBuilder;
+
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.ConstantScoreQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.TermInSetQuery;
+import org.apache.lucene.search.TermQuery;
+import org.apache.lucene.util.BytesRef;
+
+public class LuceneTextTermsPredicate extends AbstractLuceneLeafSingleFieldPredicate {
+
+	private LuceneTextTermsPredicate(Builder builder) {
+		super( builder );
+	}
+
+	public static class Factory<F>
+			extends
+			AbstractLuceneCodecAwareSearchValueFieldQueryElementFactory<TermsPredicateBuilder, F, LuceneStandardFieldCodec<F, String>> {
+		public Factory(LuceneStandardFieldCodec<F, String> codec) {
+			super( codec );
+		}
+
+		@Override
+		public Builder<F> create(LuceneSearchContext searchContext, LuceneSearchValueFieldContext<F> field) {
+			return new Builder<>( codec, searchContext, field );
+		}
+	}
+
+	private static class Builder<F> extends AbstractBuilder<F> implements TermsPredicateBuilder {
+		private final LuceneStandardFieldCodec<F, String> codec;
+
+		private String term;
+		private List<String> terms;
+		private boolean allMatch;
+
+		private Builder(LuceneStandardFieldCodec<F, String> codec, LuceneSearchContext searchContext,
+				LuceneSearchValueFieldContext<F> field) {
+			super( searchContext, field );
+			this.codec = codec;
+		}
+
+		@Override
+		public void matchingAny(Collection<?> terms, ValueConvert convert) {
+			allMatch = false;
+			fillTerms( terms, convert );
+		}
+
+		@Override
+		public void matchingAll(Collection<?> terms, ValueConvert convert) {
+			allMatch = true;
+			fillTerms( terms, convert );
+		}
+
+		@Override
+		public SearchPredicate build() {
+			return new LuceneTextTermsPredicate( this );
+		}
+
+		@Override
+		protected Query buildQuery() {
+			if ( term != null ) {
+				return new TermQuery( new Term( absoluteFieldPath, term ) );
+			}
+
+			if ( !allMatch ) {
+				List<BytesRef> bytesRefs = terms.stream().map( BytesRef::new ).collect( Collectors.toList() );
+				return new TermInSetQuery( absoluteFieldPath, bytesRefs );
+			}
+
+			BooleanQuery.Builder builder = new BooleanQuery.Builder();
+			for ( String termItem : terms ) {
+				Query query = new TermQuery( new Term( absoluteFieldPath, termItem ) );
+				builder.add( query, BooleanClause.Occur.MUST );
+			}
+			return new ConstantScoreQuery( builder.build() );
```

Comment:
Same remark as for the Elasticsearch predicate:

> So if someone calls `constantScore()` in the DSL... we will wrap the query in a constant score query twice?
> 
> Could you do something similar to `org.hibernate.search.backend.lucene.types.predicate.impl.LuceneExistsPredicate.AbstractBuilder#AbstractBuilder` instead?
> 
> Additionally, you could also add a safeguard in `applyConstantScore()` that calls a new `isAlwaysConstant()` method, so that it doesn't wrap the query unnecessarily.